### PR TITLE
Tables as lists of tuples

### LINKED ISSUE #1284: Add support for typed tables / dataframes
https://twitter.com/ShriramKMurthi/status/1462607300124979203

### PR DESCRIPTION
**Pull request by 7h3kk1d**
We have tables at home (lists of labeled tuples).

### Features

- Projection using `.` broadcasts across lists of labeled tuples
- A set of new builtin functions that have custom statics

| Builtin Name      | Description                                                                 |
|-----------------------|-----------------------------------------------------------------------------|
| `primitive_pivot` | Groups elements by label and creates labeled tuple with entries s for each group.        |
| `melt`                 | Converts a labeled tuple into a list of labeled entries.                   |
| `from_entries`    | Reconstructs a labeled tuple from a list of labeled entries.               |
| `project_labels`  | Extracts specific labels from a labeled tuple.                             |
| `select_labels`   | Extracts specific labels and creates labeled tuples for them.              |
| `omit_labels`      | Removes specific labels from a labeled tuple.                              |
| `drop_labels`     | Removes all labels from a labeled tuple, leaving only the values.          |

- Tuple extension using `...` syntax

### Remaining Issues

- [x] Builtin operations with deferral are marking labels as unexpected. DeferredAp statics needs to be modified to work like Ap
- [x] Test builtins in cases where a singleton tuple would be returned
- [x] Precedence of `...` and `ap` needs to be adjusted
- [x] Projection of duplicate label should be stuck
- [x] Projection with single quoted empty label
- [x] Test single quote labels in types
- [x] ~~Add test for backup_text in slides~~ Will finish in https://github.com/hazelgrove/hazel/pull/1840
- [x] Add empty labels, duplicate labels for from_entries, other edge cases to tuple docs
- [x] Move humidity to second position in tuple example
- [x] Add single quoted labels in examples and fix parsing and projection.

### CODE REVIEW

File: src/language/dynamics/transition/PatternMatch.re [hunk 1]
```diff
@@ -14,6 +14,7 @@ let combine_result = (r1: match_result, r2: match_result): match_result =>
 
 let rec matches = (capture, dp: Pat.t, d: DHExp.t): match_result => {
   let matches = matches(capture);
+
```
Comment (7h3kk1d) [inline src/language/dynamics/transition/PatternMatch.re]:
```suggestion
```

File: src/language/dynamics/transition/PatternMatch.re [hunk 2]
```diff
@@ -51,6 +52,7 @@ let rec matches = (capture, dp: Pat.t, d: DHExp.t): match_result => {
     matches(x, x');
   | Tuple(ps) =>
     let* ds = Unboxing.unbox(Tuple(List.length(ps)), d);
+
```
Comment (7h3kk1d) [inline src/language/dynamics/transition/PatternMatch.re]:
```suggestion
```

File: src/language/dynamics/transition/Transition.re [hunk 3]
```diff
@@ -252,6 +253,7 @@ module Transition = (EV: EV_MODE) => {
       let.wrap_closure _ = env;
       let {matches, closures} = matches(dp, d1');
       let.match env' = (env, matches, env.call_stack);
+
```
Comment (7h3kk1d) [inline src/language/dynamics/transition/Transition.re]:
```suggestion
```

File: src/language/dynamics/transition/Transition.re [hunk 4]
```diff
@@ -365,6 +367,7 @@ module Transition = (EV: EV_MODE) => {
       and. d2' =
         req_final(req(state, env), d2 => Ap2(dir, d1, d2) |> wrap_ctx, d2);
       let-unbox unboxed_fun = (Fun, d1');
+
```
Comment (7h3kk1d) [inline src/language/dynamics/transition/Transition.re]:
```suggestion
```

File: src/language/dynamics/transition/Transition.re [hunk 5]
```diff
@@ -449,6 +450,7 @@ module Transition = (EV: EV_MODE) => {
         };
       | FunNoEnv(_) => Indet
       | BuiltinFun(ident) =>
+        // print_endline("BuiltinFun");
```
Comment (7h3kk1d) [inline src/language/dynamics/transition/Transition.re]:
```suggestion
```

File: src/language/dynamics/transition/Transition.re [hunk 6]
```diff
@@ -629,32 +631,66 @@ module Transition = (EV: EV_MODE) => {
...
-      | (TupLabel(_, d), Label(name)) =>
-        LabeledTuple.has_same_labels(
-          Exp.match_tup_label(d1'),
-          Some((name, d)),
-        )
-          ? Step({
-              expr: d,
+      print_endline("Projection");
+      print_endline(DHExp.show(d1'));
+      print_endline(DHExp.show(d2'));
```
Comment (7h3kk1d) [inline src/language/dynamics/transition/Transition.re]:
```suggestion
```

File: src/language/dynamics/transition/Unboxing.re
Comment (7h3kk1d) [inline src/language/dynamics/transition/Unboxing.re:1]: Revisit what's happening in unboxing

File: src/language/statics/Info.re [hunk 7]
```diff
@@ -241,7 +243,7 @@ type label_inference('a) =
 type exp = {
   term: Exp.t, /* The term under consideration */
   ancestors, /* Ascending list of containing term ids */
-  ctx: Ctx.t, /* Typing context for the term */
+  ctx: [@show.opaque] Ctx.t, /* Typing context for the term */
```
Comment (7h3kk1d) [inline src/language/statics/Info.re]: Remove opaque

File: src/language/builtins/BuiltinsTupleOperations.re [hunk 8]
```diff
@@ -0,0 +1,268 @@
+module Fresh = IdTagged.FreshGrammar;
+open Fresh.Typ;
+open BuiltinsUtil;
+open Util;
+let builtins: list(BuiltinsUtil.fn) = [
+  {
+    name: "primitive_pivot",
```
Comment (cyrus-) [inline src/language/builtins/BuiltinsTupleOperations.re]: group_by_label

File: src/language/builtins/BuiltinsTupleOperations.re [hunk 9]
```diff
@@ -0,0 +1,268 @@
...
+          )
+          |> Option.map(Exp.to_tuple);
+
+        | _ => None
+        }
+      }),
+    custom_statics: Some(PrimitivePivot),
+  },
+  {
+    name: "melt",
```
Comment (cyrus-) [inline src/language/builtins/BuiltinsTupleOperations.re]: to_lvs

File: src/language/builtins/BuiltinsTupleOperations.re [hunk 10]
```diff
@@ -0,0 +1,268 @@
...
+              )
+            ),
+          entries,
+        );
+      Some(IdTagged.FreshGrammar.Exp.list_lit(unpivoted_entries));
+    },
+    custom_statics: Some(Ctx.Melt),
+  },
+  {
+    name: "from_entries",
```
Comment (cyrus-) [inline src/language/builtins/BuiltinsTupleOperations.re]: from_lvs

File: src/language/builtins/BuiltinsTupleOperations.re [hunk 11]
```diff
@@ -0,0 +1,268 @@
...
+            );
+          Some(Exp.to_tuple(entries));
+        | None => None
+        };
+      };
+    },
+    custom_statics: Some(OmitLabels),
+  },
+  {
+    name: "drop_labels",
```
Comment (cyrus-) [inline src/language/builtins/BuiltinsTupleOperations.re]: omit_all_labels

**Review by cyrus- — DISMISSED** [review]
